jwql/jwql_monitors/generate_preview_images.py

```diff
@@ -496,35 +497,151 @@ def find_data_channel(detectors):
     return channel
 
 
+def get_base_output_name(filename_dict):
+    """Returns the base output name used for preview images and
+    thumbnails.
+
+    Parameters
+    ----------
+    filename_dict : dict
+        A dictionary containing parsed filename parts via
+        ``filename_parser``
+
+    Returns
+    -------
+    base_output_name : str
+        The base output name, e.g. ``jw96090001002_03101_00001_nrca2_rate``
+    """
+
+    base_output_name = 'jw{}{}{}_{}{}{}_{}_'.format(
+        filename_dict['program_id'], filename_dict['observation'],
+        filename_dict['visit'], filename_dict['visit_group'],
+        filename_dict['parallel_seq_id'], filename_dict['activity'],
+        filename_dict['exposure_id'])
+
+    return base_output_name
+
+
 @log_fail
 @log_info
 def generate_preview_images():
-    """The main function of the ``generate_preview_image`` module."""
+    """The main function of the ``generate_preview_image`` module.
+    See module docstring for further details."""
 
     # Begin logging
     logging.info("Beginning the script run")
 
-    filesystem = get_config()['filesystem']
-    preview_image_filesystem = get_config()['preview_image_filesystem']
-    thumbnail_filesystem = get_config()['thumbnail_filesystem']
+    # Process programs in parallel
+    program_list = [os.path.basename(item) for item in glob.glob(os.path.join(get_config()['filesystem'], '*'))]
```

Comment:
I like the idea of breaking it up by program and creating a separate thread for each one.